File: day2/main1.py
with open("input") as f:
    max_color = {'red':12, 'green':13, 'blue':14}
    ans = 0
    lines = [x.rstrip() for x in f.readlines()]
    for idx, line in enumerate(lines):
        subsets = [x.split(',') for x in line.split(':')[1].split(';')]
        valid = True
        for subset in subsets:
            for color in subset:
                cpt = int(color.split()[0])
                c = color.split()[1]
                if cpt>max_color[c]:
                    valid = False
        if valid :
            ans += idx+1
    print(ans)

File: day2/main1.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
  unordered_map<string, int> max_color =
  {
    {"red", 12},
    {"green", 13},
    {"blue", 14}
  };

  ifstream ifs("input");
  string line;
  int ans=0;
  while(getline(ifs, line)){
    stringstream ssline(line);
    string game, semicolon, subset;
    int n;
    ssline >> game >> n >> semicolon;
    bool accepted = true;
    while(getline(ssline, subset, ';')){
      stringstream ss(subset);
      string color;
      while(getline(ss, color, ',')){
        string c;
        int cpt;
        stringstream sscolor(color);
        sscolor >> cpt >> c;
        if(cpt>max_color[c])
          accepted = false;
      }
    }
    if(accepted){
      ans+=n;
    }
  }
  cout << ans << endl;
}
